hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1439,6 +1440,238 @@ private class Emit(
+
+          infoDGEQRFResult := Code.invokeScalaObject[Int, Int, Long, Int, Long, Long, Int, Int](LAPACK.getClass, "dgeqrf",
+            M.toI,
+            N.toI,
+            ndPType.data.pType.elementOffset(aAddressDGEQRF, aNumElements.toI, 0),
+            LDA.toI,
+            tauPType.elementOffset(tauAddress, K.toI, 0),
+            workAddress,
+            LWORK
+          ),
+          Code.invokeStatic[Memory, Long, Unit]("free", workAddress.load()),
+          infoDGEQRFErrorTest
+        )
+
+        val result = if (mode == "raw") {
+          // Raw returns the column major form of A as is.
+          val rawPType = x.pType.asInstanceOf[PTuple]
+          val rawOutputSrvb = new StagedRegionValueBuilder(mb, x.pType, region)
+          val hPType = rawPType.types(0).asInstanceOf[PNDArray]
+          val tauPType = rawPType.types(1).asInstanceOf[PNDArray]
+
+          val hShapeArray = Array(N, M)
+          val hShapeBuilder = hPType.makeShapeBuilder(hShapeArray)
+          val hStridesBuilder = hPType.makeDefaultStridesBuilder(hShapeArray, mb)
+
+          val tauShapeBuilder = tauPType.makeShapeBuilder(Array(K))
+          val tauStridesBuilder = tauPType.makeDefaultStridesBuilder(Array(K), mb)
+
+          val h = hPType.construct(0, 0, hShapeBuilder, hStridesBuilder, aAddressDGEQRF, mb)
+          val tau = tauPType.construct(0, 0, tauShapeBuilder, tauStridesBuilder, tauAddress, mb)
+
+          val computeRaw = Code(
+            rawOutputSrvb.start(),
+            rawOutputSrvb.addIRIntermediate(hPType)(h),
+            rawOutputSrvb.advance(),
+            rawOutputSrvb.addIRIntermediate(tauPType)(tau),
+            rawOutputSrvb.advance(),
+            rawOutputSrvb.end()
+          )
+
+          Code(
+            alwaysNeeded,
+            computeRaw
+          )
+        }
+        else {
+          val currRow = mb.newField[Int]
+          val currCol = mb.newField[Int]
+
+          //This block assumes rDataAddress is a row major ndarray.
+          val zeroOutCorner =
+            Code.forLoop(currRow := 0, currRow < M.toI, currRow := currRow + 1,
+              Code.forLoop(currCol := 0, currCol < N.toI, currCol := currCol + 1,
+                (currRow > currCol).orEmpty(
+                  Region.storeDouble(
+                    ndPType.data.pType.elementOffset(rDataAddress, aNumElements.toI, currRow * N.toI + currCol),
+                    0.0)
+                )
+              )
+            )
+
+          val (rPType, rShapeArray) = if (mode == "r") {
+            (x.pType.asInstanceOf[PNDArray], Array(K, N))
+          } else if (mode == "complete") {
+            (x.pType.asInstanceOf[PTuple].types(1).asInstanceOf[PNDArray], Array(M, N))
+          } else if (mode == "reduced") {
+            (x.pType.asInstanceOf[PTuple].types(1).asInstanceOf[PNDArray], Array(K, N))
+          } else {
+            throw new HailException(s"Unsupported QR mode $mode")
```

Comment:
semantically this is really an assert, because it should've been caught by type checking. I think we should change to `assert(false, s"Unsupported QR mode $mode")`

Reply:
I did this because I thought I needed an `else` case for this if statement since i have to either return something or throw an exception. I could just make the `else if (mode == "reduced")` into an `else`, but I think it reads better when you can see which mode you're working with explicitly.

Reply:
I'm generally a big fan of asserts and checking all the cases and I think it's good that you chose to write it this way.